service/src/event/service.rs

```diff
@@ -130,76 +177,35 @@ impl CeramicEventService {
+    pub(crate) async fn insert_events(
+        &self,
         items: &[ReconItem<EventId>],
-    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        source: DeliverableRequirement,
+    ) -> Result<InsertResult> {
         let mut parsed_events = Vec::with_capacity(items.len());
-        let mut invalid_events = Vec::new();
+        let mut invalid = Vec::with_capacity(items.len() / 4); // assume most of the events are valid
         for event in items {
-            match Self::parse_discovered_event(event).await {
+            match CeramicEventService::parse_discovered_event(event).await {
                 Ok(insertable) => parsed_events.push(insertable),
-                Err(err) => invalid_events.push(InvalidItem::InvalidFormat {
+                Err(err) => invalid.push(InvalidItem::InvalidFormat {
                     key: event.key.clone(),
                     reason: err.to_string(),
                 }),
             }
         }
 
-        // Group events by their type
-        let mut valid_events = Vec::with_capacity(parsed_events.len());
-        let mut signed_events = Vec::with_capacity(parsed_events.len());
-        let mut time_events = Vec::with_capacity(parsed_events.len());
-        for event in parsed_events {
-            match event.event() {
-                Event::Time(_) => {
-                    time_events.push(event);
-                }
-                Event::Signed(_) => {
-                    signed_events.push(event);
-                }
-                Event::Unsigned(_) => {
-                    // Unsigned events need no extra validation.
-                    valid_events.push(event);
-                }
-            }
-        }
-
-        let (
-            (valid_signed_events, invalid_signed_events),
-            (valid_time_events, invalid_time_events),
-        ) = try_join!(
-            Self::validate_signed_events(signed_events),
-            Self::validate_time_events(time_events)
-        )?;
-
-        valid_events.extend(valid_signed_events);
-        valid_events.extend(valid_time_events);
-        invalid_events.extend(invalid_signed_events);
-        invalid_events.extend(invalid_time_events);
-
-        Ok((valid_events, invalid_events))
-    }
-
-    pub(crate) async fn insert_events(
-        &self,
-        items: &[ReconItem<EventId>],
-        source: DeliverableRequirement,
-    ) -> Result<InsertResult> {
-        let (to_insert, mut invalid) = Self::validate_events(items).await?;
+        let history_req = source.into();
+        let pending_to_insert = self.remove_pending(&parsed_events);
+        let to_validate = parsed_events.into_iter().chain(pending_to_insert).collect();
+
+        let validator = EventValidator::new(&self.pool, history_req);
+        let (to_insert, pending) = validator.validate_events(to_validate, &mut invalid).await?;
+        pending.iter().for_each(|p| {
+            invalid.push(InvalidItem::RequiresHistory {
```

Comment:
doesn't whether pending events are considered invalid or not depend on the `DeliverableRequirement` in use?  It looks like below we are still inserting events with missing history so long as we're in Lazy or Asap mode.

Reply:
Yeah, this is weird and needs adjustment. The deliverable requirement is how I got the "history requirement" enum.. I wanted something else but it may not be different enough to be useful. Right now, we turn these into invalid items and the into API/Recon result turns them into an error/pending count increment.

Pending is only supposed to ever happen for recon events, we don't want to pend migration data (just get it over with asap) and API events should have everything on the node and be rejected. Pending is different than missing history for IOD but I need to figure out how to use/share/encapsulate things across validation and IOD more nicely. IOD requires complete history, while data event validation requires only the init event, which we expect to find _soon_ if we don't already have it, so `pending`. But if we have no init event, no need to try ordering. 

Right now, I have a bit of a mix of doing the same work in both places and not knowing anything about the other or state of the event. I think tracking the validation process as a new type (e.g. including the init cid when it's used), might make it easier to reason about and be useful for the IOD stuff too.